src/Parsers/Lexer.cpp

```diff
@@ -105,44 +105,57 @@ Token Lexer::nextTokenImpl()
             if (prev_significant_token_type == TokenType::Dot)
             {
                 ++pos;
-                while (pos < end && isNumericASCII(*pos))
+                while (pos < end && (isNumericASCII(*pos) || isNumberSeparator(false, false, pos, end)))
                     ++pos;
             }
             else
             {
+                bool start_of_block = false;
                 /// 0x, 0b
                 bool hex = false;
-                if (pos + 2 < end && *pos == '0' && (pos[1] == 'x' || pos[1] == 'b' || pos[1] == 'X' || pos[1] == 'B'))
+                if (pos + 2 < end && *pos == '0' && (pos[1] == 'x' || pos[1] == 'b' || pos[1] == 'X' || pos[1] == 'B') && pos[2] != ';')
```

Comment:
Why do we check `;` ?

Reply:
Actually, I changed this logic a bit in the last of the commits. On master we had the following behaviour:
```sh
clickhouse client -q "select 0x"   # results in 0
clickhouse client -q "select 0x;"  # results in 0
clickhouse client -q "select 0x ;" # results in 0
# and same for 0b
```

That `0x` or `0b` doesn't really look like a valid number, so I changed it to the following behaviour:
```sh
clickhouse client -q "select 0x"   # results in UNKNOWN_IDENTIFIER
clickhouse client -q "select 0x;"  # results in UNKNOWN_IDENTIFIER
clickhouse client -q "select 0x ;" # results in UNKNOWN_IDENTIFIER
# and same for 0b
```

There's one more weird case, but behaviour on my branch is consistent with master:
```sql
select -_0; -- UNKNOWN_IDENTIFIER
select +_0; -- SYNTAX_ERROR
```